console/collections/src/kary_merkle_tree/mod.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod helpers;
+pub use helpers::*;
+
+mod path;
+pub use path::*;
+
+#[cfg(test)]
+mod tests;
+
+use snarkvm_console_types::prelude::*;
+
+use aleo_std::prelude::*;
+
+#[derive(Clone)]
+pub struct KaryMerkleTree<LH: LeafHash<Hash = PH::Hash>, PH: PathHash, const DEPTH: u8, const ARITY: u8> {
+    /// The leaf hasher for the Merkle tree.
+    leaf_hasher: LH,
+    /// The path hasher for the Merkle tree.
+    path_hasher: PH,
+    /// The computed root of the full Merkle tree.
+    root: PH::Hash,
+    /// The internal hashes, from root to hashed leaves, of the full Merkle tree.
+    tree: Vec<PH::Hash>,
+    /// The canonical empty hash.
+    empty_hash: PH::Hash,
+    /// The number of hashed leaves in the tree.
+    number_of_leaves: usize,
+}
+
+/// Returns the next power of `n` that's greater than or equal to `base`.
+/// Returns `None` for edge cases or in case of overflow.
+fn checked_next_power_of_n(base: usize, n: usize) -> Option<usize> {
+    if n <= 1 {
+        return None;
+    }
+
+    let mut value = 1;
+    while value < base {
+        value = value.checked_mul(n)?;
+    }
+    Some(value)
+}
+
+impl<LH: LeafHash<Hash = PH::Hash>, PH: PathHash, const DEPTH: u8, const ARITY: u8>
+    KaryMerkleTree<LH, PH, DEPTH, ARITY>
+{
+    #[inline]
+    /// Initializes a new Merkle tree with the given leaves.
```

Comment:
nit: the official recommendation is to put the attributes after the doc comments

Reply:
Resolved below